# **Caso de estudio - arboles de decision**

Una FINTECH quiere la implementacion de un modelo de preaprobacion de creditos utilizando un modelo de arbol de decision. Las variables de entrada y salida se descrbiben a continuacion:
* Preaprobacion (salida-referencia):Indica en una etapa muy temprana del credito si le presto o no le prsto a una persona
* Edad (entrada): Indica ola edad que posee una persona al momneto de solicitar un credito


In [ ]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier #modelo arbol de decision
from sklearn.metrics import confusion_matrix
#Carga automatica de las bases de datos
from google.colab import files
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
nxl= '/content/drive/MyDrive/Colab Notebooks/6. SolicitantesCrédito(USD) (1).xlsx'
XDB=pd.read_excel(nxl, sheet_name=1) #Base de datos AAA
XDB=XDB.dropna() # Elimina celdas erroneas o N/A
XDB=XDB [['Edad', 'Perscargo', 'Ingresos', 'Egresos', 'Monto (EAD)', 'PreApr']]
#Separar variables de entrada y variables de salida
XD=XDB[['Edad', 'Perscargo', 'Ingresos', 'Egresos', 'Monto (EAD)']] #Variables entrada
yd=XDB['PreApr'] #Variables de Referencia-salida





3. Se procede con la implemntacion del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD,yd)

#Para la grafica del arbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data # Corrected import statement
dot_data=export_graphviz(mar,feature_names=XD.columns)
graph=graph_from_dot_data(dot_data)
graph.write_png('tree_clasae.png')

True

4. Evaluamos el comportamineto del modelo fretne a los datos presentados

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd, ydp)
print(cm)
VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]
print('VN=',VN,'FP=',FP,'FN=',FN,'VP=',VP)

Ex=(VN+VP)/(VN+FP+FN+VP)
print('Exactitud=',Ex) #comportamiento general del modelo

Ter=(FP+FN)/(VN+FP+FN+VP)
print('Tasa de error=',Ter) #comportamiento erroneo del modelo
print('Tasa de falso positivo=',Ter)

Sen=VP/(VP+FN)
print('Sensibilidad=',Sen) #comportamiento positivo del modelo

Esp=VN/(VN+FP)
print('Especificidad=',Esp) #comportamiento frente a los prenegativos del modelo

Prec=VP/(VP+FP)
print('Precisión=',Prec) #comportamiento solamente contando los positivos

PredNeg=VN/(VN+FN)
print('Predictividad frente a los negativos=',PreN)#Comportamiento solo contando con los negativose

[[2301  658]
 [ 644 2239]]
VN= 2301 FP= 658 FN= 644 VP= 2239
Exactitud= 0.7771311194796303
Tasa de error= 0.22286888052036974
Tasa de falso positivo= 0.22286888052036974
Sensibilidad= 0.7766215747485259
Especificidad= 0.7776275768840825


## **ANALISIS DE RESULTADO**

deacuerdo con los resultados obtenidos por el modelo podemos observar de manera general que el modelo logro una exactitud que s eubico por encima del 75% en promedio (0.777131), lo que indica el buen comportamiento general del modelo frente a la preaprobacion de creditos en una FINTECH. Este valor esa por encima de los porcentajes que establece la superfinanciera para este tipo de modelos. De esta misma manera se puede observar que tanto la sensiblidad como la especificidad lograron valores ponr encima de 75% en promedio, sin embargo se destaca que la especificidad logra un porcentaje(0.777627) estuvo por encima de la sensibilidad, lo que indica que el modelo tiene la leve tendencia a prenegar creditos.
Se destaca en el arbol que las variables que tuvieron una mayor relevacoin fueron las variables Monto (EAD), seguida de los ingresos. Se destaca que el modelo arbol de decision solo logro un nodo puro el cual esta definido por la siguiente regla del negocio: Monto<=322 and ingre>376 and Egr < 685 and ingr=644 en donde la decision se denota "[10,0]", lo que indica la tendencia de esra regla a prenegar creditos con una certeza del 100%